superset/assets/src/SqlLab/components/LimitControl.jsx

```diff
@@ -0,0 +1,122 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+  Button,
+  Label,
+  FormGroup,
+  FormControl,
+  Overlay,
+  Popover,
+} from 'react-bootstrap';
+import ControlHeader from '../../explore/components/ControlHeader';
+import { t } from '../../locales';
+
+const propTypes = {
+  value: PropTypes.number,
+  defaultQueryLimit: PropTypes.number.isRequired,
+  maxRow: PropTypes.number.isRequired,
+  onChange: PropTypes.func.isRequired,
+};
+
+export default class LimitControl extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    const { value, defaultQueryLimit } = props;
+    this.state = {
+      textValue: value.toString() || defaultQueryLimit.toString(),
+      showOverlay: false,
+    };
+  }
+
+  setValueAndClose(val) {
+    this.setState({ textValue: val }, this.submitAndClose.bind(this));
```

Comment:
do the binding in constructor so it won't create new function every time this gets called.